vllm/worker/model_runner.py

```diff
@@ -39,40 +48,80 @@
-    lora_mapping: Optional[LoRAMapping]
-    lora_requests: Set[LoRARequest]
-    multi_modal_kwargs: Dict[str, torch.Tensor]
-    slot_mapping: torch.Tensor
-    num_prefill_tokens: int
-    num_decode_tokens: int
-    num_prefills: int
+@dataclass(frozen=True)
+class ModelInputForGPU(ModelRunnerInputBase):
+    """
+    This base class contains metadata needed for the base model forward pass
+    but not metadata for possible additional steps, e.g., sampling. Model
+    runners that run additional steps should subclass this method to add
+    additional fields.
+    """
+    input_tokens: Optional[torch.Tensor] = None
+    input_positions: Optional[torch.Tensor] = None
+    seq_lens: Optional[List[int]] = None
+    query_lens: Optional[List[int]] = None
+    lora_mapping: Optional["LoRAMapping"] = None
+    lora_requests: Optional[Set[LoRARequest]] = None
+    attn_metadata: Optional["AttentionMetadata"] = None
+    multi_modal_kwargs: Optional[Dict[str, torch.Tensor]] = None
+
+    def as_broadcastable_tensor_dict(
+            self) -> Dict[str, Union[int, torch.Tensor]]:
+        tensor_dict = self._get_attrs([
+            "input_tokens",
+            "input_positions",
+            "lora_requests",
+            "lora_mapping",
+            "multi_modal_kwargs",
+        ])
+        _add_attn_metadata_broadcastable_dict(tensor_dict, self.attn_metadata)
+        return tensor_dict
 
     @classmethod
-    def empty(cls, device):
-        return ModelInput(
-            input_tokens=torch.empty(0, device=device),
-            input_positions=torch.empty(0, device=device),
-            attn_metadata=None,
-            seq_lens=[],
-            query_lens=[],
-            lora_mapping=None,
-            lora_requests=set(),
-            multi_modal_kwargs={},
-            slot_mapping=torch.empty(0, device=device),
-            num_prefill_tokens=0,
-            num_decode_tokens=0,
-            num_prefills=0,
-        )
+    def new(cls: Type[TModelInputForGPU],
+            attn_backend: Optional["AttentionBackend"] = None,
+            **kwargs) -> TModelInputForGPU:
+        if attn_backend is not None:
+            kwargs = _init_attn_metadata_from_kwargs(attn_backend, **kwargs)
+        kwargs = _filter_valid_kwargs(cls, kwargs)
+        return cls(**kwargs)
 
 
-class ModelRunner:
+@dataclass(frozen=True)
+class ModelInputForGPUWithSamplingMetadata(ModelInputForGPU):
+    """
+    Used by the ModelRunner.
+    """
+    sampling_metadata: Optional["SamplingMetadata"] = None
+    # Used for speculative decoding. We do not broadcast it because it is only
+    # used by the driver worker.
+    is_prompt: Optional[bool] = None
+
+    def as_broadcastable_tensor_dict(
+            self) -> Dict[str, Union[int, torch.Tensor]]:
+        tensor_dict = super().as_broadcastable_tensor_dict()
+        _add_sampling_metadata_broadcastable_dict(tensor_dict,
+                                                  self.sampling_metadata)
```

Comment:
inheritance does not give any benefit here, but rather decreases the readability. same as above comments, we can just expand the code, write out these fields directly.

Reply:
The benefit of inheritance here is that if the broadcasted fields change, we only need to modify them in one place. Otherwise, a developer can mistakenly add it in only one place, type checking will not complain, but the broadcast will be buggy for all subclasses.

Do you prefer to duplicate the code? I disagree in this case but I don't feel strongly about it.

Reply:
I think duplicating the code here is fine. When someone modify them in one place, the worker side will have a code to consume that field. Although type checkers cannot ccomplain, test code will catch it.